disttask/framework/scheduler/scheduler.go

```diff
@@ -0,0 +1,284 @@
+	if err != nil {
+		s.onError(err)
+		return s.getError()
+	}
+
+	for {
+		subtask, err := s.subtaskTable.GetSubtaskInStates(s.id, task.ID, proto.TaskStatePending)
+		if err != nil {
+			s.onError(err)
+			break
+		}
+		if subtask == nil {
+			logutil.Logger(s.logCtx).Info("scheduler finished subtasks", zap.Any("step", task.Step))
+			break
+		}
+		s.updateSubtaskState(subtask.ID, proto.TaskStateRunning)
+		if err := s.getError(); err != nil {
+			break
+		}
+
+		minimalTasks := scheduler.SplitSubtask(subtask.Meta)
+		logutil.Logger(s.logCtx).Info("split subTask", zap.Any("cnt", len(minimalTasks)), zap.Any("subtask_id", subtask.ID))
+		for _, minimalTask := range minimalTasks {
+			minimalTaskWg.Add(1)
+			j := minimalTask
+			minimalTaskCh <- func() {
+				s.runMinimalTask(runCtx, j, task.Type, task.Step)
+				minimalTaskWg.Done()
+			}
+		}
+		minimalTaskWg.Wait()
+		if err := s.getError(); err != nil {
+			if errors.Cause(err) == context.Canceled {
+				s.updateSubtaskState(subtask.ID, proto.TaskStateCanceled)
+			} else {
+				s.updateSubtaskState(subtask.ID, proto.TaskStateFailed)
+			}
+			break
+		}
+		s.updateSubtaskState(subtask.ID, proto.TaskStateSucceed)
+	}
+
+	return s.getError()
+}
+
+func (s *InternalSchedulerImpl) runMinimalTask(minimalTaskCtx context.Context, minimalTask proto.MinimalTask, tp string, step int64) {
+	logutil.Logger(s.logCtx).Info("scheduler run a minimalTask", zap.Any("step", step), zap.Any("minimal_task", minimalTask))
+	select {
+	case <-minimalTaskCtx.Done():
+		s.onError(minimalTaskCtx.Err())
+		return
+	default:
+	}
+	if s.getError() != nil {
+		return
+	}
+
+	executor, err := createSubtaskExecutor(minimalTask, tp, step)
+	if err != nil {
+		s.onError(err)
+		return
+	}
+
+	if err = executor.Run(minimalTaskCtx); err != nil {
+		s.onError(err)
+	}
+}
+
+// Rollback rollbacks the scheduler task.
+func (s *InternalSchedulerImpl) Rollback(ctx context.Context, task *proto.Task) error {
+	rollbackCtx, rollbackCancel := context.WithCancel(ctx)
+	defer rollbackCancel()
+	s.registerCancelFunc(rollbackCancel)
+
+	s.resetError()
+	logutil.Logger(s.logCtx).Info("scheduler rollback a step", zap.Any("step", task.Step))
+	scheduler, err := createScheduler(task)
+	if err != nil {
+		s.onError(err)
+		return s.getError()
+	}
+	subtask, err := s.subtaskTable.GetSubtaskInStates(s.id, task.ID, proto.TaskStateRevertPending)
+	if err != nil {
+		s.onError(err)
+		return s.getError()
+	}
+	if subtask == nil {
+		logutil.BgLogger().Warn("scheduler rollback a step, but no subtask in revert_pending state", zap.Any("step", task.Step))
+		return nil
+	}
+	s.updateSubtaskState(subtask.ID, proto.TaskStateReverting)
+	if err := s.getError(); err != nil {
+		return err
+	}
+
+	err = scheduler.Rollback(rollbackCtx)
+	if err != nil {
+		s.updateSubtaskState(subtask.ID, proto.TaskStateRevertFailed)
+		s.onError(err)
+	} else {
+		s.updateSubtaskState(subtask.ID, proto.TaskStateReverted)
```

Comment:
If an instances fails forever, a subtask will never be called which will cause the global task never end. Is there any TODO to handle it?

Reply:
We will add heartbeat mechanism later